### DISCUSSION TITLE
Enable Fingerprint, Smart Lock, and Trust Agents in emulated devices

### DISCUSSION THREAD

**Discussion by bhein-sherpa6** [Genymobile/feedback#59] [Genymotion Desktop] [answered]
Hey dev team,

I work on testing a Mobile Device Management (MDM) solution for Android devices, and one of the initiatives is to control device policies and settings. In order to test this, we need to be able to validate that the MDM is properly controlling most/all Android settings (i.e. the settings in the com.android.settings app). We have been doing this through emulated devices in Android Studio AVD, but have been experiencing flakiness and slow performance with that solution, so we purchased a Genymotion Desktop license, which appears to be much more stable and consistent.

However, when we went in to look at the following settings, they were not available at all:

- Fingerprint
- Smart lock
- Trust agents

Is there any particular reason that your solution could not support these, just as Android's AVD does?

I'm sure there are many others that we haven't encountered yet that will be missing, so more generally, is there any pushback to supporting most/all of these settings as they work in Android AVD emulators?

Thank you for your consideration!

Brandon

**Comment by guillaume-roche** [comment, marked as answer]
Hi Brandon,

Thanks a lot for your interest in Genymotion, your inputs and feedbacks are really valuable.

First, let me say that bringing new features on all our supported Android versions and making them work in all flavors of Genymotion is really a tough challenge. Also, some features need additional design to ensure they are available in a way that makes sense in an Android emulator.

However, I can tell you that fingerprint support is in our roadmap. We are doing our best to have it released by the end of the year.

Regarding Smart lock and Trust agents, those feature require the Google Play Services, and they are not available in the AOSP. You can try the following and see if it suits your needs:

- Install the GApps in your device and reboot
- Set a PIN code in Parameters / Security 
- Still in the security page, you should be able to open the Trust agents and enable Smart lock.
- Then the Smart Lock menu is available

https://github.com/Genymobile/feedback/assets/6976869/67731d2d-2b15-4c6f-9e40-5cb9554c4424

I hope this will help you moving forward with Genymotion. Don't hesitate if you have further details about your use case.

Best regards,
Guillaume

**Comment by bhein-sherpa6** [comment]
Hello, thank you for your response.

“First, let me say that bringing new features on all our supported Android versions and making them work in all flavors of Genymotion is really a tough challenge. Also, some features need additional design to ensure they are available in a way that makes sense in an Android emulator.”

Trust me, I completely understand. We’ve spent many days sorting through the differences between each android version, and each os image within those versions as well.

“Regarding Smart lock and Trust agents, those feature require the Google Play Services, and they are not available in the AOSP”

Alright that makes sense, so you’re using a custom os image without Google Play services enabled then. I will attempt to add it after the device boots, but ultimately if this could be set programmatically (either through adb/gmtool/etc.) that would be optimal, since we’re implementing fully-automated tests.

Fingerprint support will be excellent. We are currently able to set fingerprints through adb (“emu finger touch xyz”) so if this (or with gmtool) is still possible with your approach that would be appreciated.

Overall, we are very much enjoying your Desktop solution over the Android Studio AVD from a stability perspective. We are seeing far fewer phone crashes/freezes, and our automation scripts are running much smoother with your solution.

Thank you for your information!

**Reply by bhein-sherpa6** [reply]
Ok, I’ve followed your steps and I am now seeing the settings in question as suggested. As a follow up, would it be possible to offer a device profile with this enabled by default, instead of having to download it after the device is created?

Again, the motivation here is that automation scripts should be able to create and start a device at runtime (using gmtool) and have most/all of the configuration already set up.

Thank you for your consideration!

**Reply by bhein-sherpa6** [reply]
I followed this guide: https://support.genymotion.com/hc/en-us/articles/4909301061533-How-to-install-the-Open-GAPPs-package-manually- in order to programmatically do this, however, this still means that I have to manually download the Gapps archive, store this somewhere, and use adb to install. To reiterate, if you could consider implementing a way to either:

a.) Provide a device profile by default which includes these Gapps. This will obviously significantly increase the size of the download for these devices, but I think think the payoff is worth it. This would be preferred.
b.) Provide a way to trigger your "Open Gapps" toolbar widget programmatically. This could be done through gmtool, or through your Java API, which already controls other toolbar feature such as GPS and Battery.

Sorry for the spam, I'm actively working through the optimal approach to automate all of this. Thanks again!

**Reply by guillaume-roche** [reply]
Hello again,

I agree that  shipping Genymotion with built-in Gapps would be handy for our users. Unfortunately, we are not allowed to do so for legal reasons. 

A solution that I see would be to manually provision a reference virtual device with everything setup correctly. Then you can clone the reference device to a temporary device, and use it for your testing. This way, the reference device is not impacted by your testing.

Regarding automation in a broader perspective, we are focusing our efforts on Genymotion SaaS and Genymotion Device Image, as they are more suitable to be used in CI/CD environments than Genymotion Desktop. Genymotion Desktop is more appropriate for non-automated use cases, like manual testing, pentesting, app live demos or UX/UI experiments.

Best regards,
Guillaume

**Reply by bhein-sherpa6** [reply]
"I agree that shipping Genymotion with built-in Gapps would be handy for our users. Unfortunately, we are not allowed to do so for legal reasons."

That is indeed unfortunate. Ok, thank you so much for your information and prompt reply. I'm extremely grateful for your attention. I won't take any more of your time on this, I think I have what I need to continue using the Genymotion Desktop solution for now. We will consider migrating to your other solutions at some point if it becomes an issue.

**Comment by agavignet** [comment]
Hi @bhein-sherpa6 

Regarding the fingerprint, Desktop 3.6 released today brings a new Biometrics widget, you can see a quick demo on our blog post https://www.genymotion.com/blog/genymotion-desktop-3-6/ 

Don't hesitate to give us feedbacks ;)
Adrien
